openpype/pipeline/create/context.py

```diff
@@ -177,6 +177,127 @@ def prepare_failed_creator_operation_info(
     }
 
 
+class ChangedItem(object):
+    def __init__(self, old_value, new_value):
+        self._old_value = copy.deepcopy(old_value)
+        self._new_value = copy.deepcopy(new_value)
+        self._changed = self._old_value != self._new_value
+
+        old_is_dict = isinstance(old_value, dict)
+        new_is_dict = isinstance(new_value, dict)
+        children = {}
+        changed_keys = set()
+        available_keys = set()
+        new_keys = set()
+        old_keys = set()
+        if old_is_dict and new_is_dict:
+            old_keys = set(old_value.keys())
+            new_keys = set(new_value.keys())
+            available_keys = old_keys | new_keys
+            for key in available_keys:
+                item = ChangedItem(
+                    old_value.get(key), new_value.get(key)
+                )
+                children[key] = item
+                if item.changed or key not in old_keys or key not in new_keys:
+                    changed_keys.add(key)
+
+        elif old_is_dict:
+            old_keys = set(old_value.keys())
+            available_keys = set(old_keys)
+            changed_keys = set(available_keys)
+            for key in available_keys:
+                children[key] = ChangedItem(old_value.get(key), None)
+
+        elif new_is_dict:
+            new_keys = set(new_value.keys())
+            available_keys = set(new_keys)
+            changed_keys = set(available_keys)
+            for key in available_keys:
+                children[key] = ChangedItem(None, new_value.get(key))
+
+        self._changed_keys = changed_keys
+        self._available_keys = available_keys
+        self._children = children
+        self._old_is_dict = old_is_dict
+        self._new_is_dict = new_is_dict
+        self._old_keys = old_keys
+        self._new_keys = new_keys
+
+    def __getitem__(self, key):
+        return self._children[key]
+
+    def __bool__(self):
+        return self._changed
+
+    def __iter__(self):
+        for key in self.changed_keys:
+            yield key
+
+    def get(self, key, default=None):
+        return self._children.get(key, default)
+
+    def keys(self):
+        return self.changed_keys
+
+    def items(self):
+        if self.is_dict:
+            yield None, self.changes
+        else:
+            for item in self.changes.items():
```

Comment:
`self.changes` would return `(self.old_value, self.new_value)` here? That wouldn't have `.items()`?

Reply:
It would return `{key: (old, value)}` if the item is dictionary.

Reply:
The code says:
```python
    @property
    def changes(self):
        if not self.is_dict:
            return (self.old_value, self.new_value)
        ...
```

So whenever we're in this `else` branch here we're in that _if not_ branch there, no?

Reply:
I had switched logic here in `items`.